Concept - How GenAI and LLMs work (#40428)
dotnet/docs: docs/ai/conceptual/how-genai-and-llms-work.md

---
title: "How Generative AI and LLMs work"

> description: "Understand how Generative AI and large language models (LLMs) work and how they may be useful in your .NET projects."

Reviewer: ```suggestion
description: "Understand how Generative AI and large language models (LLMs) work and how they might be useful in your .NET projects."
```

author: haywoodsloan
ms.topic: concept-article
ms.date: 04/04/2024

#customer intent: As a .NET developer, I want to understand how Generative AI and large language models (LLMs) work and how they may be useful in my .NET projects.

---

> # How Generative AI and LLMs work

Reviewer: ```suggestion
# How generative AI and LLMs work
```

> Generative AI is a type of artificial intelligence capable of creating original content, such as natural language, images, audio, code, etc. The output of a generative AI is based on the inputs provided by the user. One common way for users to interact with generative AI is through chat applications which use natural language as their input. ChatGPT, developed by OpenAI, is a popular example of this. Generative AI applications which use natural language as an input are powered by large language models (LLMs) to perform natural language processing (NLP).

Reviewer: ```suggestion
Generative AI is a type of artificial intelligence capable of creating original content, such as natural language, images, audio, and code. The output of a generative AI is based on the inputs provided by the user. One common way for users to interact with generative AI is through chat applications that use natural language as their input. ChatGPT, developed by OpenAI, is a popular example of this. Generative AI applications that use natural language as an input are powered by large language models (LLMs) to perform natural language processing (NLP).
```

> ## How does Generative AI work?

Reviewer: ```suggestion
## How generative AI works
```

> All generative AI is built on top of models. These models are trained with large sets of data in the form of content, such as natural language, images, audio, code, etc. Generative AI models use the patterns identified in the training data to produce new, statistically similar content.

Reviewer: ```suggestion
All generative AI is built on top of models. These models are trained with large sets of data in the form of content, such as natural language, images, audio, and code. Generative AI models use the patterns identified in the training data to produce new, statistically similar content.
```

> The input provided by the user is the seed that the generative AI model builds its output from. The input is first parsed into a form of data that the model can understand. The model then uses that data to identify matching patterns from its training that it will combine to build the final output. Generative AI models are designed to produce unique content, so they won't generate the same output for identical inputs.

Reviewer: ```suggestion
The input provided by the user is the seed that the generative AI model builds its output from. The input is first parsed into a form of data that the model can understand. The model then uses that data to identify matching patterns from its training that it combines to build the final output. Generative AI models are designed to produce unique content, so they won't generate the same output for identical inputs.
```

> Generative AI applications which support natural language as an input or output utilize LLMs to do so. The LLM is used to perform NLP, classifying the input text and determining its sentiment. That classification and sentiment analysis is used by the generative AI model to identify patterns and build the output. If the output is text, the LLM alone can be used to generate it. If the ouput is audio, images, etc., additional models are used to provide the data and patterns for generating outputs in that format.

Reviewer: ```suggestion
Generative AI applications that support natural language as an input or output utilize LLMs to do so. The LLM is used to perform NLP, which classifies the input text and determines its sentiment. That classification and sentiment analysis is used by the generative AI model to identify patterns and build the output. If the output is text, the LLM alone can be used to generate it. If the output is audio or images, additional models are used to provide the data and patterns for generating outputs in that format.
```

> ## Potential uses of Generative AI

Reviewer: ```suggestion
## Potential uses of generative AI
```

Because generative AI applications can support a variety of output formats, there are many potential use cases for them.

### Natural language generation

Some generative AI applications can produce natural language outputs, such as:

- Ideas for healthy meals
- Professionally worded summaries for a LinkedIn profile
- Rich descriptions of items in an online storefront

> For example, a user may input the request, "Give me three ideas for a healthy breakfast including peppers." A generative AI application might produce the following output:

Reviewer: ```suggestion
For example, given the user input "Give me three ideas for a healthy breakfast including peppers", a generative AI application might produce the following output:
```

_OK, here are three healthy breakfasts that include peppers:_

1. _Egg and Pepper Scramble: Dice up some bell peppers and sauté them in a pan until they're slightly softened. Then, add in some beaten eggs and scramble everything together. Serve with whole-grain toast for a nutritious breakfast._

2. _Pepper and Avocado Toast: Toast a slice of whole-grain bread and top it with sliced bell peppers, mashed avocado, and a sprinkle of salt and pepper. This is a great way to get some healthy fats from the avocado along with the vitamin C from the peppers._

3. _Breakfast Burrito with Peppers: Sauté some diced bell peppers, onions, and scrambled eggs together in a pan. Then, spoon the mixture onto a whole-grain tortilla and top with some salsa or hot sauce. Roll it up and enjoy a satisfying and nutritious breakfast._

### Image generation

> Some generative AI applications can produce image outputs from natural language inputs. Depending on the application and its model(s), users may be able to specify:

Reviewer: ```suggestion
Some generative AI applications can produce image outputs from natural language inputs. Depending on the application and its models, users might be able to specify:
```

- The people or things to appear the image
- The artistic style to create the image in
- References for generating similar images

Image generation can be used to create virtual avatars for online accounts, design logos for a business, or provide artistic inspiration for creators. For example, a user may input the request, "Create an image of an elephant eating a burger." A generative AI application might produce the following output:

:::image type="content" source="../media/how-genai-and-llms-work/generated-image.png" lightbox="../media/how-genai-and-llms-work/generated-image.png" alt-text="Example AI generated image of an elephant eating a hamburger.":::

### Audio generation

> Some generative AI applications can produce audio outputs from natural language inputs. Depending on the application and its model(s), users may be able to:

Reviewer: ```suggestion
Some generative AI applications can produce audio outputs from natural language inputs. Depending on the application and its models, users might be able to:
```

- Synthesize natural sounding voices from input text
- Create music in a specific style or featuring certain instruments
- Modify input audio files based on a set criteria provided in natural language

Audio generation can be used to provide spoken responses from digital voice assistants, add backing instruments to songs for music production, or reproduce a user's original voice from reference recordings if they have experienced some form of speech impairment.

### Code generation

> Some generative AI applications can produce code from natural language inputs. Depending on the application and its model(s), users may be able to generate code:

Reviewer: ```suggestion
Some generative AI applications can produce code from natural language inputs. Depending on the application and its models, users might be able to generate code that:
```

> - In the language of their choice

Reviewer: ```suggestion
- Is in the language of their choice
```

- Create music in a specific style or featuring certain instruments
- Modify input audio files based on a set criteria provided in natural language

Audio generation can be used to provide spoken responses from digital voice assistants, add backing instruments to songs for music production, or reproduce a user's original voice from reference recordings if they have experienced some form of speech impairment.

### Code generation

Some generative AI applications can produce code from natural language inputs. Depending on the application and its model(s), users may be able to generate code:

- In the language of their choice
- That accepts a specific input and output structure
- That uses specific technologies based on their reference documentation

Code generation can be used to provide quick implementations of common operations, such as sorting or filtering data, speed up translating code from one language to another, or answer questions about a technology based on specifed reference documentation. For example, a user may input the request, "Show me how to code a game of tic-tac-toe with Python". A generative AI application might produce the following output:

> _Here's a simple example of how you can code a game of tic-tac-toe in C#:_

Author: Good suggestion! I did just that and omitted the bulk of the code sample with a comment mentioning that.
